File: java/src/isel/leic/pc/monitors/CounterSemaphoreSN.java
/*--------------------------------------------------------------
 * counter semaphore implementation with execution delegation
 * and specific notification, using a ReentrantLock with multiple conditions
 */

package isel.leic.pc.monitors;

import isel.leic.pc.utils.TimeoutHolder;


// We use our LinkedList that returns the node references for added items,
// in order to optimize node remove implementation
import isel.leic.pc.utils.LinkedList;

import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.ReentrantLock;


public class CounterSemaphoreSN {
    private ReentrantLock monitor;
    private int permits;

    private static class Request {
        final int n;
        boolean granted;
        Condition condition;
        Request(int n, Condition cond) {
            this.n = n; condition = cond;
        }
    }

    private LinkedList<Request> waiters;

    public CounterSemaphoreSN(int initialPermits) {
        if (initialPermits >=0)
            permits = initialPermits;
        monitor = new ReentrantLock();
        waiters = new LinkedList<>();
    }

    private void notifyWaiters() {
        while (waiters.size() > 0 && permits >= waiters.getFirst().n) {
            Request req = waiters.removeFirst();
            req.granted = true;
            permits-= req.n;
            req.condition.signal();
        }
    }

    public boolean acquire(int n, long timeout) throws InterruptedException {
        synchronized(monitor) {
            // Here is the correct test to do.
            // On the lecture, the test for empty waiters list was forgotten,
            // which permits barging, that is, is possible that an
            // acquiring passes in front of waiting threads, jeopardizing FIFO discipline!
            if (permits >= n && waiters.size() == 0) {
                permits -= n;
                return true;
            }
            if (timeout == 0) return false;
            TimeoutHolder th = new TimeoutHolder(timeout);
            Request req = new Request(n, monitor.newCondition());
            LinkedList.Node node = waiters.add(req);

            try {
                do {
                    req.condition.await(th.value(), TimeUnit.MILLISECONDS);
                    // on execution delegation pattern, just check n was granted
                    if (req.granted)
                        return true;

                    if (th.timeout() ) {
                        waiters.remove(node);
                        notifyWaiters();
                        return false;
                    }
                }
                while (true);
            }
            catch(InterruptedException e) {
                if (req.granted) {
                    // on execution delegation pattern, in case of n was granted
                    // just delay the interrupt and return success
                    Thread.currentThread().interrupt();
                    return true;
                }
                waiters.remove(node);
                notifyWaiters();
                throw e;
            }
        }
    }

    public void release(int n) {
        synchronized (monitor) {
            permits += n;
            notifyWaiters();
        }
    }

}

File: dotnet/Aula-2018-10-09/CounterSemaphoreSN.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using SynchUtils;

namespace Aula_2018_10_09 {
    /// <summary>
    /// Semaphore implementation using execution delegation
    /// and specific notification to minimize context switches
    /// </summary>
    public class CounterSemaphoreSN {
        private Object monitor;
        private int permits;
        LinkedList<Request> waiters;

        private class Request {
            internal int reqUnits;
            internal bool granted;

            internal Request(int reqUnits) {
                this.reqUnits = reqUnits;
            }
        }

        public CounterSemaphoreSN(int initialPermits) {
            if (initialPermits < 0)
                throw new InvalidOperationException();

            permits = initialPermits;
            monitor = new Object();
            waiters = new LinkedList<Request>();
        }

        private void notifyWaiters() {
            while (waiters.Count > 0) {
                LinkedListNode<Request> n = waiters.First;
                if (n.Value.reqUnits > permits) return;
                permits -= n.Value.reqUnits;
                n.Value.granted = true;
                waiters.RemoveFirst();
                //thewaiters node is the condition!
                monitor.Notify(n);
            }
             
        }

        public bool Acquire(int units, int timeout) { // throws InterruptedException  
            lock (monitor) {
                // non blocking path
                if (permits >= units && waiters.Count == 0) {
                    permits -= units;
                    return true;
                }
                if (timeout == 0)
                    return false;
                // blocking path
                // prepare wait
                TimeoutHolder th = new TimeoutHolder(timeout);
                var node = waiters.AddLast(new Request(units));
              
                try {
                    do {
                        // use waiters node as a new condition!
                        monitor.Wait(node, th.Value);
                        if (node.Value.granted) {
                            return true;
                        }
                        if (th.Timeout) {
                            waiters.Remove(node);
                            notifyWaiters();
                            return false;
                        }
                    }
                    while (true);
                }
                catch (ThreadInterruptedException e) {
                    if (node.Value.granted) {
                        Thread.CurrentThread.Interrupt();
                        return true;
                    }
                    waiters.Remove(node);
                    notifyWaiters();
                    throw e;
                }
            }
        }

        public void Release(int releases) {
            lock (monitor) {
                permits += releases;
                notifyWaiters();
            }
        }
    }
}
